File: Sources/Sentry/SentryFileIOTrackingIntegration.m
#import "SentryFileIOTrackingIntegration.h"
#import "SentrySwift.h"

@interface SentryFileIOTrackingIntegration ()

@property (nonatomic, strong) SentryFileIOTracker *tracker;

@end

@implementation SentryFileIOTrackingIntegration

- (BOOL)installWithOptions:(SentryOptions *)options
{
    if (![super installWithOptions:options]) {
        return NO;
    }

    self.tracker = [[SentryDependencyContainer sharedInstance] fileIOTracker];
    [self.tracker enable];

    [SentryDependencyContainer.sharedInstance.nsDataSwizzling startWithOptions:options
                                                                       tracker:self.tracker];
    [SentryDependencyContainer.sharedInstance.nsFileManagerSwizzling startWithOptions:options
                                                                              tracker:self.tracker];

    return YES;
}

- (SentryIntegrationOption)integrationOptions
{
    return kIntegrationOptionIsTracingEnabled | kIntegrationOptionEnableAutoPerformanceTracing
        | kIntegrationOptionEnableFileIOTracing;
}

- (void)uninstall
{
    [self.tracker disable];

    [SentryDependencyContainer.sharedInstance.nsDataSwizzling stop];
    [SentryDependencyContainer.sharedInstance.nsFileManagerSwizzling stop];
}

@end

File: Sources/Swift/Integrations/Performance/IO/SentryFileIOTrackingIntegration.swift
@_implementationOnly import _SentryPrivate

typealias FileIOTrackingIntegrationProvider = FileIOTrackerProvider & NSDataSwizzlingProvider & NSFileManagerSwizzlingProvider

final class SentryFileIOTrackingIntegration<Dependencies: FileIOTrackingIntegrationProvider>: NSObject, SwiftIntegration {
    private let tracker: SentryFileIOTracker
    private let nsDataSwizzling: SentryNSDataSwizzling
    private let nsFileManagerSwizzling: SentryNSFileManagerSwizzling

    init?(with options: Options, dependencies: Dependencies) {
        // Check if tracing is enabled
        guard options.isTracingEnabled else {
            SentrySDKLog.debug("Not going to enable \(Self.name) because isTracingEnabled is disabled.")
            return nil
        }

        // Check if auto performance tracing is enabled
        guard options.enableAutoPerformanceTracing else {
            SentrySDKLog.debug("Not going to enable \(Self.name) because enableAutoPerformanceTracing is disabled.")
            return nil
        }

        // Check if file IO tracing is enabled
        guard options.enableFileIOTracing else {
            SentrySDKLog.debug("Not going to enable \(Self.name) because enableFileIOTracing is disabled.")
            return nil
        }

        self.tracker = dependencies.fileIOTracker
        self.nsDataSwizzling = dependencies.nsDataSwizzling
        self.nsFileManagerSwizzling = dependencies.nsFileManagerSwizzling

        super.init()

        tracker.enable()

        nsDataSwizzling.start(withOptions: options, tracker: tracker)
        nsFileManagerSwizzling.start(withOptions: options, tracker: tracker)
    }

    func uninstall() {
        tracker.disable()

        nsDataSwizzling.stop()
        nsFileManagerSwizzling.stop()
    }

    static var name: String {
        "SentryFileIOTrackingIntegration"
    }
}
